lib/remap-functions/src/join.rs

```diff
@@ -0,0 +1,156 @@
+use remap::prelude::*;
+use std::borrow::Cow;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Join;
+
+impl Function for Join {
+    fn identifier(&self) -> &'static str {
+        "join"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Array(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let separator = arguments.optional("separator").map(Expr::boxed);
+
+        Ok(Box::new(JoinFn { value, separator }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct JoinFn {
+    value: Box<dyn Expression>,
+    separator: Option<Box<dyn Expression>>,
+}
+
+impl Expression for JoinFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let array = self.value.execute(state, object)?.try_array()?;
+
+        let string_vec = array
+            .iter()
+            .map(|s| s.try_bytes_utf8_lossy().map_err(Into::into))
+            .collect::<Result<Vec<Cow<'_, str>>>>()
+            .map_err(|_| "all array items must be strings")?;
+
+        let separator: String = self
+            .separator
+            .as_ref()
+            .map(|s| {
+                s.execute(state, object)
+                    .and_then(|v| Value::try_bytes(v).map_err(Into::into))
+            })
+            .transpose()?
+            .map(|s| String::from_utf8_lossy(&s).to_string())
+            .unwrap_or_else(|| "".into());
+
+        let joined = string_vec.join(&separator);
+
+        Ok(Value::from(joined))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        self.value
+            .type_def(state)
+            // Always fallible because the `value` array could contain non-strings
+            .into_fallible(true)
```

Comment:
An array has an `inner_type_def` which indicates the type of the elements of that array. So you could check this and only be fallible if the inner_type_def is not `Value::Bytes`.

Reply:
Also TIL for me!